(Performance) ManagedEntities - Only update an entity if its declaration has changed

totten opened the pull request:
Overview
----------------------------------------

Generally, the purpose of `ManagedEntities` is to keep a database-record up-to-date with the description in code.

This optimization should improve the performance of a typical system-flush/cache-clear. For example, on my local `drupal-clean` build (*with ~65 managed entities*), the typical runtime of `cv flush` drops from 5.1s to 1.6s. 

The benefits should be more significant for systems with more managed-entities.

Before
----------------------------------------

To reconcile, we walk through _all the entities_ and issue API calls (`create`, `update`, `delete`) to ensure that the database is up-to-date with the declaration.

However, during most flushes, there has been no real change in either the entity-declaration or the live-entity. Consequently, the `update` (usually) isn't doing anything useful.

After
----------------------------------------

We only perform an `update` if the entity-declaration has changed in some way.

Comments
----------------------------------------

* While working on #32339 with the wmff build, I noticed that `rebuildMenuAndCaches()` was slow -- and that the configuration declared *almost 1,000 managed-entities*. That's a huge number -- much more than `ManagedEntities` was originally designed for. This kind of configuration should see bigger gains.
* I haven't done very deep testing. I wouldn't be surprised if there are some test-failures. I've only done the quick benchmark.
* This does change the interpretation of [the `update` policy](https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_managed/):

    | Policy | Old Interpretation | New Interpretation |
    | -- | -- | -- | 
    |  `always` | Re-save whenever there is a system-flush | Re-save whenever there is a change in the canonical declaration |
    | `unmodified` | Re-save whenever there is a system-flush _and the user has not edited the live record_. | Re-save whenever there is a change in the canonical declaration _and the user has not edited the live record_. |
    | `never` | Never try to re-save. Lock-in the original value. | (*Unchanged*) |

    The new interpretation still follows the same *configuration precedence* (e.g. `always` prioritizes canonical-declarations; `never` prioritizes local-configuration) -- but it applies enforcement less frequently.

* What if you want strict enforcement -- i.e. your intent is to *strictly prevent* edits to the managed-entity? 
    * Well, `update=>always` was the closest to doing that, but it was never strict. (You might say it was 66% of the way there.)
    * With this formulation, it still gives precedence, but it's even less strict. (You might say it's only 33% of the way there.)
    * To get this use-case to 100%, we probably need some kind of _read-only_ flag on the record -- whenever someone tries to edit the live entity, we first ensure that it is not protected by a read-only flag.

* ~~On a gut-level, I expect there will be some scenarios where one wants to forcibly re-save. But I'm not exactly sure where that ends-up. An advanced flush-option? Something you do for full core-upgrades... but not for intermediate `cv flush`es? I'm not sure For the moment, there's just a discreet PHP flag as a nod in that direction.~~ (*This has been re-worked. It does a full re-save whenever you upgrade the system, and it does a targeted re-save when you (re)enable an extension.*)

totten committed:
CRM_Utils_Array - Add deepSort()

totten committed:
Managed - Declare new column `checksum`

totten committed:
ManagedEntities - Only update entity if its declaration has changed (variant C)

colemanw commented on CRM/Upgrade/Incremental/php/SixTwo.php:
```diff
@@ -29,6 +29,8 @@ class CRM_Upgrade_Incremental_php_SixTwo extends CRM_Upgrade_Incremental_Base {
   public function upgrade_6_2_alpha1($rev): void {
     $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
+    $this->addTask('Add column "civicrm_managed.checksum"', 'addColumn', 'civicrm_managed',
+      'checksum', "varchar(45) NULL COMMENT 'Configuration of the managed-entity when last stored'");
```
Fun fact: we have a new helper for this: `alterSchemaField`. (you can see it in action 2 lines below this one)

colemanw commented on CRM/Core/ManagedEntities.php:
```diff
@@ -135,10 +135,15 @@ private function backfillDefaults(array &$item): void {
+   *   If FALSE, then only re-save entities when the declaration changes.
    */
-  private function reconcileEntities(array $plan): void {
+  private function reconcileEntities(array $plan, bool $force = FALSE): void {
```
Non-blocker:

This new `$force` param looks to be deeply internal; it's not even reachable when calling the `public function reconcile()`. If it's useful, then maybe it should be exposed there and also as a param in the `Api4.Managed.reconcile` action.
Or if it's not useful, then maybe it should be removed per YAGNI?

totten replied:
This was one of the questions raised in the description.

After looking at the tests, it seemed apparent that we need to re-save whenever an extension is toggled (i.e. propagating `is_active`). 

I also think that doing it during a system-upgrade (but not a manual flush) strikes a balance between (a) not needing to re-save all time and (b) still giving some enforcement to `update=>always`.

Anyway, the parameter is replaced by a computation now.

totten commented on CRM/Core/ManagedEntities.php:
```diff
@@ -188,6 +188,11 @@ private function filterPlanByChanges(array $plan): array {
       }
 
+      // (4) Some items have magic weights which shift up+down.
+      if (CoreUtil::isType($item['entity_type'], 'SortableEntity')) {
```
@colemanw I haven't really seen much about "SortableEntity", so I'm not really sure what to do with this.

* On one hand,  there's `testManagedNavigationWeights()`. The test was initially failing -- but I added this extra rule to make it pass.
* On the other hand, a huge number of entities are "sortable". This rule means that lots of entities are re-saved on the theory that their `weight` might need to be recomputed.
    * This negates a lot of the performance benefit. Comparing the output of `time cv flush`, I see:
        * 5.5s (on master, where all entities are re-saved)
        * 1.6s (if we skip re-saves for all entities -- before 9a59ab13b626e98503f75a8824d01a9a0deaf628)
        * 5.0s (if we still re-save sortable entities -- after 9a59ab13b626e98503f75a8824d01a9a0deaf628)

Is there some other mechanism that could reconcile weights (*perhaps a single operation*) or some other guard that could tell us if weights need to change?

I guess... maybe a rule like this?

* If the overall `$plan` shows that all `Navigation` mgds are unchanged (*matching checksums*), then we can skip all `update`s for `Navigation` mgds.
* If the overall `$plan` has any changes (inserts/deletes/new-checksums) involving `Navigation` mgds, then we must re-save all `Navigation` mgds.

That's the best I can see. But I really have to look at `SortableEntity` as a blackbox. Maybe there's some better way...

totten replied:
OK, well, I still don't fully grok `SortableEntity` lifecycle, but I think that's not actually the problem that I was having with `testManagedNavigationWeights()`.

Rather, it was a mismatch between (a) how the test mocked activation/deactivation of modules and (b) how the new code detected activation/deactivation of modules.

I think the next push should pass the test without needing this heavy-handed commit.

totten committed:
CiviCase - During mgd reconciliation, ensure new cases+activities propagate

The problem becomes apparent in some of the mixin tests.

Historically, they were asymptomatic because `ManagedEntities::reconcile()`
would cast a very wide net -- and some of the other items in that net would
_coincidentally_ flush caches.

However, if the reconciliation is more targetted, then we don't get
a timely flush.

totten commented:
So far, new test-run is looking better for all scenarios that previously failed. Removing the "Draft" flag.

ufundo commented:
Really really excited for this @totten !

> To get this use-case to 100%, we probably need some kind of read-only flag on the record -- whenever someone tries to edit the live entity, we first ensure that it is not protected by a read-only flag.

Could we use `is_reserved` is for this? Use the old policy where canonical data has `'is_reserved' => TRUE`? I'm not sure if that maps exactly onto what we want - but given that it would be retaining the old behaviour in those cases it seems fairly safe?

colemanw commented:
This all looks great & tests are passing. I think the conversation about `is_reserved` is helpful but non-blocking.

totten commented:
> (*Description*)... wmff build... declared almost 1,000 managed-entities...

Just following-up about that example. I tried the patch on a local copy (*and manually adjusted schema version to match it*). The time required for `cv flush` dropped by 90% -- from 1.5 minutes to 10 seconds.

> (@ufundo) Could we use is_reserved is for this? Use the old policy where canonical data has 'is_reserved' => TRUE? I'm not sure if that maps exactly onto what we want - but given that it would be retaining the old behaviour in those cases it seems fairly safe?

I suppose there are a couple variations:

* __ `is_reserved` as an existing pattern shared by ~20 entities__: : I suppose we would need a stronger meta-definition of what it means.  There is a long-tail of adapting to more entities (`CustomField`, `SavedSearch`, `SearchDisplay`, `Navigation`, etc).
* __`is_reserved` as a new property on `civicrm_managed`__: It feels like a generic event-listener (`hook_pre` or `civi.api.authorize` or somesuch) could get big coverage and decent strictness with minimal code. However, when it comes to fine-tuning the UI (*e.g. greyed-out buttons and boxes*), then there would still be a long-tail.

> (@colemanw) This all looks great & tests are passing. I think the conversation about is_reserved is helpful but non-blocking.

Woot!

totten commented on CRM/Core/ManagedEntities.php:
```diff
@@ -77,6 +77,16 @@ public function reconcile($modules = NULL) {
     $declarations = $this->getDeclarations($modules);
     $plan = $this->createPlan($declarations, $modules);
+    if (!CRM_Core_Config::isUpgradeMode()) {
+      $plan = $this->optimizePlan($plan);
```
Note to self: The optimization here is related to the new test-failures on `cv`'s `ExtensionLifecycleTest`. It took a while to track/confirm. To reproduce, I have to run two-part test on `cv`, eg

```bash
./scripts/run-tests.sh --debug tests/Command/ExtensionLifecycleTest.php --filter '(testLifecycleWithShortNames|testDownloadWithExplicitUrl)'
```

(*I'm a bit tired from hunting it down -- but that gives something better to focus on...*)

angusfretwell commented:
This is excellent, thanks for this @totten!
